File: junction/samples/serverless/http/follower_service.cc
#include <boost/algorithm/string/join.hpp>
#include <string>
#include <unordered_map>
#include <vector>

#include "gateway_client.h"
#include "lib/httplib.h"
#include "watchdog.h"

namespace {
bool enable_interception = false;

std::unordered_map<int, std::vector<int>> GenerateFollowers(int count) {
  std::unordered_map<int, std::vector<int>> db;
  db.reserve(count);

  for (int i = 0; i < count; ++i) {
    std::vector<int> users;
    for (int j = 0; j < i; j++) { users.push_back(j); }
    db[i] = users;
  }
  return db;
}

std::unordered_map<int, std::vector<int>> followers_db = GenerateFollowers(100);

void GetFollowersHandler(const httplib::Request &req, httplib::Response &res) {
  try {
    int user_id = std::stoi(req.path_params.at("id"));
    const std::vector<int> &followers = followers_db.at(user_id);
    std::string req_path = "/user/";
    std::vector<std::string> names;
    for (const int &id : followers) {
      names.push_back(
          CallGateway(req_path + std::to_string(id), enable_interception));
    }
    res.set_content(boost::algorithm::join(names, ", "), "text/plain");
  } catch (...) {
    res.status = httplib::StatusCode::NotFound_404;
    res.set_content("User does exist", "text/plain");
  }
}
}  // namespace

int main(int argc, char *argv[]) {
  if (argc > 1 && std::strcmp(argv[1], "--int") == 0) {
    enable_interception = true;
  }
  WatchDog w("follower", "/followers/:id", GetFollowersHandler);
  w.Run();
  return 0;
}

File: junction/samples/serverless/http/follower_service.py
import http.client
from http.server import BaseHTTPRequestHandler
import sys
from watchdog import WatchDog
import patch

patch.patch_socket()

# --- 1. CONFIGURATION ---
GATEWAY_IP = "10.10.1.1"
GATEWAY_PORT = 8080

# --- 2. GENERATE DATABASE ---
def generate_followers(count):
    db = {}
    for i in range(count):
        db[i] = list(range(i))
    return db

FOLLOWERS_DB = generate_followers(100)

# --- 3. GATEWAY CLIENT HELPER ---
def call_gateway(path):
    conn = None
    try:
        # Python tries to set TCP_NODELAY here, but our patch will catch the error
        conn = http.client.HTTPConnection(GATEWAY_IP, GATEWAY_PORT)
        
        conn.request("GET", path)
        resp = conn.getresponse()
        
        if resp.status == 200:
            return resp.read().decode('utf-8')
        return "" 

    except Exception as e:
        print(f"[Follower] Gateway call failed: {e}", file=sys.stderr)
        return ""
    finally:
        if conn:
            conn.close()

# --- 4. HANDLER LOGIC ---
class FollowerHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        prefix = "/followers/"
        if not self.path.startswith(prefix):
            self.send_error(404, "Not Found")
            return

        try:
            user_id_str = self.path[len(prefix):]
            user_id = int(user_id_str)
            
            if user_id not in FOLLOWERS_DB:
                raise ValueError("User not found")
            
            followers = FOLLOWERS_DB[user_id]
            
            names = []
            for f_id in followers:
                req_path = f"/user/{f_id}"
                name = call_gateway(req_path)
                names.append(name)
            
            response_body = ", ".join(names)
            
            self.send_response(200)
            self.send_header("Content-Type", "text/plain")
            self.send_header("Content-Length", str(len(response_body)))
            self.end_headers()
            self.wfile.write(response_body.encode('utf-8'))
            
        except (ValueError, KeyError):
            error_msg = "User does exist" 
            self.send_response(404)
            self.send_header("Content-Type", "text/plain")
            self.send_header("Content-Length", str(len(error_msg)))
            self.end_headers()
            self.wfile.write(error_msg.encode('utf-8'))

    def log_message(self, format, *args):
        return

if __name__ == "__main__":
    w = WatchDog("follower", FollowerHandler)
    w.run()
